File: Android/src/com/amay077/listdialogtest/ColorSelectPreference.java
package com.amay077.listdialogtest;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Color;
import android.preference.Preference;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

public class ColorSelectPreference extends Preference {
	final Integer[] DEFINE_COLORS = { Color.BLUE, Color.GREEN, Color.RED };
	View _widgetView;
	private int _selectedColor;
	
	static class ColorListAdapter extends ArrayAdapter<Integer> {
		public ColorListAdapter(Context context, Integer[] colors) {
			super(context, android.R.layout.simple_list_item_1, colors);
		}
		
		@Override
		public View getView(int position, View convertView, ViewGroup parent) {
			TextView view = (TextView)super.getView(position, convertView, parent);
			view.setBackgroundColor(getItem(position));
			view.setTextColor(Color.TRANSPARENT); // 色値が表示されないように隠す
			return view;
		}
	}

	public ColorSelectPreference(Context context, AttributeSet attrs) {
		super(context, attrs);
		setWidgetLayoutResource(R.layout.pref_colorselect_widget);
	}
	
	@Override
	protected void onClick() {
		ColorListAdapter adapter = new ColorListAdapter(getContext(), DEFINE_COLORS);
		new AlertDialog.Builder(getContext())
    	.setTitle(getTitle())
    	.setAdapter(adapter, new DialogInterface.OnClickListener() {
			@Override
			public void onClick(DialogInterface dialog, int which) {
				int color = DEFINE_COLORS[which];
				setColor(color);
				updateColor(color);
			}
		}).show();
	}
	
	private void setColor(int color) {
		final boolean changed = color != _selectedColor;
		_selectedColor = color;
		persistInt(color);
		
		if (changed) {
			notifyChanged();
		}
	}
	public int getColor() {
		return _selectedColor;
	}

	@Override
	protected void onBindView(View view) {
		super.onBindView(view);
		_widgetView = view.findViewById(R.id.my_widget);
		_selectedColor = getPersistedInt(-1);
		if (_selectedColor != -1) {
			updateColor(_selectedColor);
		}
	}
	
	private void updateColor(int color) {
		if (_widgetView == null) {
			return;
		}
		_widgetView.setBackgroundColor(color);
	}
}

File: Xamarin.Android/ColorSelectPreference.cs
using System;
using Android.Preferences;
using Android.Views;
using Android.Graphics;
using Android.Widget;
using Android.Content;
using Android.Util;
using Android.App;

namespace ColorPickerSample
{
    public class ColorSelectPreference : Preference
    {
        readonly Color[] DEFINE_COLORS = { Color.Blue, Color.Green, Color.Red };
        View _widgetView;

        Color _selectedColor;
        public Color Color
        {
            get { return _selectedColor; }
            set 
            {
                var changed = value != _selectedColor;
                _selectedColor = value;
                PersistInt(value.ToArgb());
                
                if (changed) 
                {
                    NotifyChanged();
                }
            }
        }

        class ColorListAdapter : ArrayAdapter<Color>
        {
            public ColorListAdapter(Context context, Color[] colors) 
                : base(context, Android.Resource.Layout.SimpleListItem1, colors)
            {
            }

            public override View GetView(int position, View convertView, ViewGroup parent)
            {
                var view = base.GetView(position, convertView, parent) as TextView;
                view.SetBackgroundColor(new Color(GetItem(position)));
                view.SetTextColor(Color.Transparent); // 色値が表示されないように隠す
                return view;
            }
        }

        public ColorSelectPreference(Context context, IAttributeSet attrs) : base(context, attrs)
        {
        this.WidgetLayoutResource = ColorPickerSample.Resource.Layout.PrefColorselectWidget;
        }
        
        protected override void OnBindView(View view)
        {
            base.OnBindView(view);
            _widgetView = view.FindViewById<View>(ColorPickerSample.Resource.Id.my_widget);
            var colorValue = GetPersistedInt(-1);
            if (colorValue != -1) {
                this.Color = new Color(colorValue);
                UpdateColor(this.Color);
            }
        }

        protected override void OnClick()
        {
            var adapter = new ColorListAdapter(this.Context, DEFINE_COLORS);
            new AlertDialog.Builder(this.Context)
                .SetTitle(this.Title)
                .SetAdapter(adapter, 
                (s, e) =>
                {
                    var color = DEFINE_COLORS [e.Which];
                    this.Color = color;
                    UpdateColor(color);
                })
                .Show();
        }
        
        private void UpdateColor(Color color) 
        {
            if (_widgetView == null) {
                return;
            }
            _widgetView.SetBackgroundColor(color);
        }    
    }
}